proposals/0724-qualified-lists.rst

```diff
@@ -0,0 +1,395 @@
+
+and it would be convenient to use list literals instead. This was even explicitly listed as a restriction in the original ``OverloadedLists`` `design <https://gitlab.haskell.org/ghc/ghc/-/wikis/overloaded-lists>`_.
+
+This proposal would desugar list literals to a build-like form instead, so that ``M.[4, "hello", True]`` desugars to
+
+::
+
+  M.buildList 3 (\cons nil -> 4 `cons` ("hello" `cons` (True `cons` nil)))
+
+For a suitable ``M.buildList``, this is enough to support heterogenous list literals: see *Section 4.6 Heterogeneous Lists*.
+
+Proposed Change Specification
+-----------------------------
+
+Introduce ``-XQualifiedLists`` that desugars literal list syntax to function calls in a similar way to ``-XQualifiedDo`` (`docs <https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/qualified_do.html>`_, `proposal <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0216-qualified-do.rst>`_).
+
+As long as the desugared expressions/patterns type check, users are free to define these functions however they want. No whitespace is allowed between the ``.`` and the module name / literal.
+
+Currently, list literals have the following desugaring:
+
+.. list-table::
+    :align: left
+
+    * - **Expression**
+      - **Enabled extensions**
+      - **Desugared expression syntax**
+    * - ``[x, y]``
+      -
+      - ``x Prelude.: y Prelude.: Prelude.[]``
+    * - ``[x, y]``
+      - ``-XOverloadedLists``
+      - ``GHC.Exts.fromListN 2 (x Prelude.: y Prelude.: Prelude.[])``
+
+With ``-XQualifiedLists``, we gain the following syntaxes:
+
+.. list-table::
+    :align: left
+
+    * - **New expression syntax**
+      - **Desugared expression syntax**
+    * - ``M.[]``
+      - ``M.buildList 0 (\cons nil -> nil)``
+    * - ``M.[x, y]``
+      - ``M.buildList 2 (\cons nil -> x `cons` (y `cons` nil))``
+    * - ``M.[x ..]``
+      - ``M.buildListEnum (Data.List.Qualified.Experimental.EnumFrom x)``
+    * - ``M.[x, y ..]``
+      - ``M.buildListEnum (Data.List.Qualified.Experimental.EnumFromThen x y)``
+    * - ``M.[x .. y]``
+      - ``M.buildListEnum (Data.List.Qualified.Experimental.EnumFromTo x y)``
+    * - ``M.[x, y .. z]``
+      - ``M.buildListEnum (Data.List.Qualified.Experimental.EnumFromThenTo x y z)``
+
+.. list-table::
+    :align: left
+
+    * - **New pattern syntax**
+      - **Desugared pattern syntax**
+    * - ``M.[x, _, y]``
+      - ``M.FromListCons x (M.FromListCons _ (M.FromListCons y M.FromListNil))``
+
+``Data.List.Qualified.Experimental`` will initially live in ``ghc-experimental``, eventually merged into ``GHC.Exts``. It will contain:
+
+::
+
+  data EnumFrom from then to
+    = EnumFrom from
+    | EnumFromThen from then
+    | EnumFromTo from to
+    | EnumFromThenTo from then to
+
+It is highly recommended that all types with ``IsList`` instances defined provide a ``.Interpolate`` module with the below definitions, to enable locally-scoped overloading over ``-XOverloadedLists``, for example:
```

Comment:
How about defining

```hs
  buildList :: IsList l => ((Item l -> Item l -> [Item l]) -> [Item l] -> [Item l]) -> MyList (Item l)
  buildList f = L.fromList (f (:) [])
```

(same for `buildListEnum`) near `IsList` and then recommend to define a local, monomorphised alias of `GHC.Exts.buildList` in the module defining the `IsList` instance? No need to duplicate this implementation.

Reply:
Good idea, done